### ISSUE TITLE
Reuse code shared between search-rg extension and vscode core

### ISSUE THREAD

**Issue by roblourens** [Microsoft/vscode#53406]
The rg SearchProvider extension uses code that's also used inside vscode. Since there is no way to directly share vscode's internal code with an extension, I need to decide whether it should be shared in a common node module between the two, or just forked.

Lots of the shared code is trivial and is ok to fork, the exception is `quickOpenScorer.ts` which is about 600 lines long.

One complication is that this code is used from 'browser' which can't import commonJS modules. So maybe I can publish it as a UMD module.

**Comment by roblourens** [comment]
I propose creating `vscode-fuzzy` (name TBD) that would contain

- src/vs/base/parts/quickopen/common/quickOpenScorer.ts
- src/vs/base/common/filters.ts
- src/vs/base/common/comparers.ts
- And their tests

I'll publish it as UMD modules, which assuming that our loader is compatible with what TS outputs as UMD, would be easy.

The other dependencies of these files are trivial. This feels like a cohesive chunk of code that makes sense in it's own repo/npm module. It represents most of the shared code between the extension and vscode, without factoring out all of src/vs/base/common.

**Comment by jrieken** [comment]
cc @alexandrudima 

> I'll publish them as UMD modules, which assuming that our loader is compatible with what TS outputs as UMD, would be easy.

We can load it but I am not so sure about bundling.

**Comment by bpasero** [comment]
For the case of `URI` I think we went a different path and simply copied the contents to a module but did not end up using it in our core. Are we proposing to start sharing code between core and extensions through UMD modules? If so, would we apply this to URI as well?

**Comment by jrieken** [comment]
Yeah, it was the same (unsolved) issue back then

**Comment by alexandrudima** [comment]
AFAIK the easiest way to share code at this time btw VS Code core (not monaco editor) and an extension would be a node module that is authored with commonjs.

**Comment by roblourens** [comment]
@jrieken so you copy changes from our uri.ts to vscode-uri every time you make a change? That's what I was trying to avoid but these files don't change that much so maybe I just leave them forked in the same way for now.

@alexandrudima I think we wouldn't be able to load commonjs modules since we use amd.

**Comment by jrieken** [comment]
> @jrieken so you copy changes from our uri.ts to vscode-uri every time you make a change? That's what I was trying to avoid but these files don't change that much so maybe I just leave them forked in the same way for now.

Yes, it's a full copy with some minor adjustment due to other dependencies. This work OK because uri doesn't change much. For *real* code re-use a different story would be better. 

>  I think we wouldn't be able to load commonjs modules since we use amd.

@roblourens No, the loader does support loading it but not bundling it. That means you cannot use them from within `/common/` and `/browser/`

**Comment by alexandrudima** [comment]
I would add that the loader also doesn't know how to load commonjs in the browser (useful for the monaco editor)
